src/device/intrinsics/random.jl

```diff
@@ -0,0 +1,49 @@
+function get_global_thread_index()
+    block_id = (gridDim().x * (blockIdx().y - 1)) + (blockIdx().x - 1)
+
+    thread_id = (block_id * (blockDim().x * blockDim().y)) +
+                ((threadIdx().y - 1) * blockDim().x) +
+                (threadIdx().x - 1)
+
+    return thread_id + 1
+end
+
+function xorshift(x::UInt32)::UInt32
+    x = xor(x, x << 13)
+    x = xor(x, x >> 17)
+    x = xor(x, x << 5)
+    return x
+end
+
+@eval @inline global_random_state() =
+    Base.llvmcall(
+        $("""@global_random_state = weak externally_initialized global i$(WORD_SIZE) 0
+             define i64 @entry() #0 {
+                 %ptr = load i$(WORD_SIZE), i$(WORD_SIZE)* @global_random_state, align 8
```

Comment:
Maybe introduce a `ThreadLocalRNG` struct that encapsulates the RNG state (so now just `val::UInt32`) to decouple the size of the RNG state from the word size? That also makes it easier to add additional data if we ever want to support more fancy RNGs.

I also don't think this is correct, since you use a `LLVMPtr{UInt32}` in `rand()` whereas the global here depends on the CPU word size.

Reply:
It's a bit of a hack, but that LLVM global actually stores a literal pointer address, mostly because I couldn't get it to work using an actual pointer type. The type of the global has no impact on the size of the RNG state (I've verified this), since again it's only an address.